pallets/pallet-bonded-coins/src/curves/mod.rs

```diff
@@ -107,46 +108,26 @@ fn calculate_accumulated_passive_issuance<Balance: Fixed>(passive_issuance: &[Ba
 
 pub(crate) fn convert_to_fixed<T: Config>(x: u128, denomination: u8) -> Result<CurveParameterTypeOf<T>, ArithmeticError>
 where
-	<CurveParameterTypeOf<T> as Fixed>::Bits: TryFrom<u128>,
+	<CurveParameterTypeOf<T> as Fixed>::Bits: TryFrom<U256>,
 {
-	let decimals = 10u128
-		.checked_pow(u32::from(denomination))
+	let decimals = U256::from(10)
+		.checked_pow(denomination.into())
 		.ok_or(ArithmeticError::Overflow)?;
-
-	// Scale down x by the denomination using integer division, truncating any
-	// fractional parts of the result for now. Will overflow if down-scaling is not
-	// sufficient to bring x to a range representable in the fixed point format.
-	// This can happen if capacity(u128) / 10^denomination > capacity(fixed).
-	let mut scaled_x: CurveParameterTypeOf<T> = x
-		.checked_div(decimals)
-		.ok_or(ArithmeticError::DivisionByZero)?
-		.checked_to_fixed()
-		.ok_or(ArithmeticError::Overflow)?;
-
-	// Next we handle the remainder of the division
-	let remainder = x.checked_rem_euclid(decimals).ok_or(ArithmeticError::DivisionByZero)?;
-
-	// If the remainder is not 0, convert to fixed, scale it down and add the
-	// resulting fractional part to the previous result
-	if remainder > 0 {
-		// Convert remainder to fixed-point and scale it down by `decimals`,
-		let fractional = remainder
-			// This can overflow if 10^denomination > capacity(fixed)
-			.checked_to_fixed::<CurveParameterTypeOf<T>>()
-			.ok_or(ArithmeticError::Overflow)?
-			// This would overflow if 10^denomination exceeds the capacity of the _underlying integer_ of the fixed
-			// (e.g., i128 for an I75F53). However, there is no denomination that is representable in a u128 but not in
-			// an i128.
-			.checked_div_int(decimals.checked_into().ok_or(ArithmeticError::Overflow)?)
-			.ok_or(ArithmeticError::DivisionByZero)?;
-
-		// Combine both parts
-		scaled_x = scaled_x
-			// Overflow is theoretically impossible as we are adding a number < 1 to a fixed point where the fractional
-			// part is 0.
-			.checked_add(fractional)
-			.ok_or(ArithmeticError::Overflow)?;
-	};
-
-	Ok(scaled_x)
+	// Convert to U256 so we have enough bits to perform lossless scaling.
+	let mut x_u256 = U256::from(x);
+	// Shift left to produce the representation that our fixed type would have (but
+	// with extra integer bits that would potentially not fit in the fixed type).
+	x_u256.shl_assign(CurveParameterTypeOf::<T>::frac_nbits());
```

Comment:
Any idea if this can panic? I didn't find a `checked_` variant of shl, so probably not?

Reply:
The macro is implementing the trait: 

```
		impl<T> $crate::core_::ops::ShlAssign<T> for $name where T: Into<$name> {
			fn shl_assign(&mut self, shift: T) {
				*self = *self << shift;
			}
		}
```

the `<<` operation can [panic](https://users.rust-lang.org/t/intentionally-overflow-on-shift/11859)

Reply:
Ok but the panic would occur if the shift is larger than the size of the number, right? Which would be an insane misconfiguration. Should we just prevent against misconfiguration the way @ntn-x2 suggested [here](https://github.com/KILTprotocol/kilt-node/pull/794#discussion_r1838198725) and not check this at runtime?

Reply:
I am against removing the checks from the runtime. Anything that the type system does not guarantee against should be checked, and having it in the runtime does not imply removing the static tests for the config itself.

Reply:
Actually as long as we don't have this configurable this cannot panic, because the U256 is large enough to work with any Fixed type that exists